Alerting: Add database table for persisting alerting configuration

gotjosh opened the pull request:
**What this PR does / why we need it**:

Alerting configuration is persisted by the API but picked up on the alerting instance via the database. We need a database table to store it.


**Special notes for your reviewer**:

This only includes the migration and a method to get configuration, I'll leave the persisting of the configuration to a separate pull request.

gotjosh committed:
Alerting: Add database table for persisting alerting configuration

gotjosh committed:
Fix the linter

papagian commented on pkg/services/ngalert/database_mig.go:
```diff
@@ -107,3 +107,19 @@ func alertInstanceMigration(mg *migrator.Migrator) {
+			{Name: "id", Type: migrator.DB_BigInt, IsPrimaryKey: true, IsAutoIncrement: true},
+			{Name: "alertmanager_configuration", Type: migrator.DB_Text, Nullable: false},
+			{Name: "alertmanager_templates", Type: migrator.DB_Text, Nullable: true},
+			{Name: "configuration_version", Type: migrator.DB_NVarchar, Length: 3, Nullable: true}, // In a format of vXX e.g. v1, v2, v10, etc
```
I don't have the full picture but I just want to make sure that it's intentionally nullable.

papagian commented on pkg/services/ngalert/database_mig.go:
```diff
@@ -107,3 +107,19 @@ func alertInstanceMigration(mg *migrator.Migrator) {
+
+func alertmanagerConfigurationMigration(mg *migrator.Migrator) {
+	alertConfiguration := migrator.Table{
+		Name: "alert_configuration",
```
I think that `alert_configuration` can be confusing. I would prefer `alertmanager_configuration` or just `alertmanager`.

gotjosh replied:
The thing is that this might not be alertmanager specific, I have the feeling we might end up including things that are not only related to the alertmanager but how we configure the "alert service" as a whole.

papagian replied:
I don't have a strong opinion, whatever works for me.

papagian commented on pkg/services/ngalert/database_mig.go:
```diff
@@ -107,3 +107,19 @@ func alertInstanceMigration(mg *migrator.Migrator) {
+		Name: "alert_configuration",
+		Columns: []*migrator.Column{
+			{Name: "id", Type: migrator.DB_BigInt, IsPrimaryKey: true, IsAutoIncrement: true},
+			{Name: "alertmanager_configuration", Type: migrator.DB_Text, Nullable: false},
+			{Name: "alertmanager_templates", Type: migrator.DB_Text, Nullable: true},
```
If the table is renamed these could be renamed also to something shorter like `configuration` and `templates `.

papagian reviewed (commented):
I have some questions/suggestions. The main is whether the receivers will be stored separately or under a JSON.

gotjosh committed:
Address review comments

papagian commented on pkg/services/ngalert/models/alertmanager.go:
```diff
@@ -0,0 +1,21 @@
+type AlertConfiguration struct {
+	ID int64 `xorm:"pk autoincr 'id'"`
+
+	AlertmanagerConfiguration string
```
I see.

papagian commented on pkg/services/ngalert/database_mig.go:
```diff
@@ -107,3 +107,19 @@ func alertInstanceMigration(mg *migrator.Migrator) {
+func alertmanagerConfigurationMigration(mg *migrator.Migrator) {
+	alertConfiguration := migrator.Table{
+		Name: "alert_configuration",
+		Columns: []*migrator.Column{
```
Configurations will be shared among all the organisations? If not, then there should be an additional column `org_id`.

gotjosh replied:
That's my current assumption, only 1 configuration for all organisations.

gotjosh committed:
Don't split templates and configuration

It is already bundled together as part of a of the API so might as well
marshall it directly.
